File: packages/system-sdk-js/src/webhooks/verifier.js
const crypto = require('crypto');

class WebhookVerifier {
    /**
     * Verify the HMAC SHA-256 signature of an incoming webhook.
     * 
     * @param {string} rawPayload The stringified JSON payload
     * @param {string} signature Header value from X-Orchestrator-Signature
     * @param {string} webhookSecret Your project's webhook secret starting with "whsec_"
     * @returns {boolean} True if signature is valid
     */
    static verifySignature(rawPayload, signature, webhookSecret) {
        if (!signature || !webhookSecret) return false;

        const expectedSignature = crypto
            .createHmac('sha256', webhookSecret)
            .update(rawPayload)
            .digest('hex');

        // Prevent timing attacks
        return crypto.timingSafeEqual(
            Buffer.from(signature),
            Buffer.from(expectedSignature)
        );
    }
}

module.exports = { WebhookVerifier };

File: packages/system-sdk-python/central_payment/webhooks/verifier.py
import hmac
import hashlib
from typing import Dict, Any, Optional
import requests

class WebhookVerifier:
    @staticmethod
    def verify_signature(raw_payload: str, signature: str, webhook_secret: str) -> bool:
        """
        Verify the HMAC SHA-256 signature of an incoming webhook.
        """
        if not signature or not webhook_secret:
            return False

        # Create HMAC-SHA256 signature
        expected_signature = hmac.new(
            webhook_secret.encode('utf-8'),
            raw_payload.encode('utf-8'),
            hashlib.sha256
        ).hexdigest()

        # Constant time comparison to prevent timing attacks
        return hmac.compare_digest(expected_signature, signature)
